arbor/backends/gpu/matrix_fine.cu

```diff
@@ -143,66 +142,67 @@ void solve_matrix_fine(
             // cells require more time steps than others, but we have to solve
             // all the matrices at the same time. When a cell finishes, we put a
             // `0` on the diagonal to mark that it should not be solved for.
-            if (d[pos]!=0) {
+            if (d[pos0]!=0) {
 
                 // each branch perform substitution
-                T factor = u[pos] / d[pos];
-                for (unsigned i=0; i<len-1; ++i) {
-                    const unsigned next_pos = pos + width;
-                    d[next_pos]   -= factor * u[pos];
-                    rhs[next_pos] -= factor * rhs[pos];
-
-                    factor = u[next_pos] / d[next_pos];
-                    pos = next_pos;
+                for (unsigned pos = pos0, pos_next = pos + width; pos < posN; pos += width, pos_next += width) {
+                    const auto u_ = u[pos];
+                    const auto d_ = d[pos];
+                    const auto d_next = d[pos_next];
+                    const auto rhs_ = rhs[pos];
+                    const auto rhs_next = rhs[pos_next];
+
+                    const T factor = -u_/d_;
+                    d[pos_next]   = fma(factor, u_, d_next);
+                    rhs[pos_next] = fma(factor, rhs_, rhs_next);
                 }
-
                 // Update d and rhs at the parent node of this branch.
                 // A parent may have more than one contributing to it, so we use
                 // atomic updates to avoid races conditions.
-                const unsigned parent_index = next_lvl_meta.matrix_data_index;
-                const unsigned p = parent_index + lvl_parents[tid];
-                //d[p]   -= factor * u[pos];
-                gpu_atomic_add(d  +p, -factor*u[pos]);
-                //rhs[p] -= factor * rhs[pos];
-                gpu_atomic_add(rhs+p, -factor*rhs[pos]);
+                const auto u_ = u[posN];
+                const auto d_ = d[posN];
+                const auto rhs_ = rhs[posN];
+                const T factor = -u[posN] / d[posN];
```

Comment:
shouldn't you use `u_` and `d_` in this expression?

Reply:
yes